Federated Hpa feature doc

irfanurrehman opened the pull request:
cc @kubernetes/sig-federation-pr-reviews @kubernetes/sig-federation-feature-requests 
cc @quinton-hoole 

![Allow edits from maintainers checkbox](https://help.github.com/assets/images/help/pull_requests/allow-maintainers-to-make-edits-sidebar-checkbox.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/5487)
<!-- Reviewable:end -->

zacharysarah commented:
@irfanurrehman :wave: I'd like to make some small edits. Would you please make sure that the box to allow edits from maintainers is checked?

irfanurrehman commented:
> @irfanurrehman  I'd like to make some small edits. Would you please make sure that the box to allow edits from maintainers is checked?

@zacharysarah apologies for missing it while raising the PR. Done now.

quinton-hoole commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[docs/tasks/administer-federation/hpa.md, line 2 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6cP4ieV8S1Ffy2fHv:-Ku6cP4ieV8S1Ffy2fHw:bwtey65) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L2)):*
> ```Markdown
> ---
> title: Federated Hpas
> ```

I'd spell it out: "Federated Horizontal Pod Autoscalers (HPA)"

---

*[docs/tasks/administer-federation/hpa.md, line 6 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6cZXZ7ZAJPl8CDq-k:-Ku6cZXZ7ZAJPl8CDq-l:b-59sdy9) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L6)):*
> ```Markdown
> 
> {% capture overview %}
> This guide explains how to use Hpas in the Federation control plane.
> ```

HPA's ?

---

*[docs/tasks/administer-federation/hpa.md, line 65 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6d8lyDnMx1VQpc-ZX:-Ku6d8lyDnMx1VQpc-ZY:b-8i7ca4) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L65)):*
> ```Markdown
> 
> Once a federated Hpa is created, the federation control plane will partition and
> create an Hpa in all underlying Kubernetes clusters.
> ```

Please mention cluster selectors to restrict the clusters into which the HPA gets federated.

---

*[docs/tasks/administer-federation/hpa.md, line 118 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6dlHCXQY6RbnIv8n2:-Ku6dlHCXQY6RbnIv8n3:b-4cho21) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L118)):*
> ```Markdown
> ## Deleting a Federated Hpa
> 
> You can delete a federated Hpa as you would delete a Kubernetes
> ```

I think it's worth mentioning that this will delete all underlying HPA's, if cascading deletion is used.

---

*[docs/tasks/administer-federation/hpa.md, line 131 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6dzROL-7xwYkv-Jyr:-Ku6dzROL-7xwYkv-Jys:b-5aby2y) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L131)):*
> ```Markdown
> 
> To a `federation user` interacting with federated control plane (or simply federation),
> the interaction is same as interacting with a normal k8s cluster (but with a limited set of
> ```

'almost identical' rather than 'the same'.

---

*[docs/tasks/administer-federation/hpa.md, line 146 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6eInOhNVJSQjqNgY8:-Ku6eInOhNVJSQjqNgY9:bs26wmr) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L146)):*
> ```Markdown
> they are needed most, or in other words where the load is more. Federated hpa feature
> achieves this by manipulating the min and max replicas on the hpas it creates in the
> federated clusters. It reduces the max and min replicas from those clusters which does
> ```

I think you need to be clearer here that the federation does not actually monitor the CPU utilization itself, but rather relies on the underlying cluster HPA's to do the auto-scaling, and only re-assigns min and max replicas between clusters in an attempt to prevent individual clusters from running up against their max replica limits.  A brief example might help to clarify this.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487)*
<!-- Sent from Reviewable.io -->

quinton-hoole commented:
I made a few suggested edits.  None of them are showstoppers, and can be addressed in a followup PR if necessary.

tengqm commented on docs/tasks/administer-federation/hpa.md:
```diff
@@ -0,0 +1,154 @@
+except in the number of min and max replicas. The federation control plane
+will ensure that the sum of max replicas in each cluster matches the specified
+max replicas on the federated hpa object and sum of min replicas will be greater
+then or equal to the min specified on the federated hpa object. The sum of min
```
then -> than

tengqm commented on docs/tasks/administer-federation/hpa.md:
```diff
@@ -0,0 +1,154 @@
+will ensure that the sum of max replicas in each cluster matches the specified
+max replicas on the federated hpa object and sum of min replicas will be greater
+then or equal to the min specified on the federated hpa object. The sum of min
+is greater or equal to specified min because an Hpa into a particular cluster
```
greater -> greater than

tengqm commented on docs/tasks/administer-federation/hpa.md:
```diff
@@ -0,0 +1,154 @@
+
+By default, first max replicas are spread equally in all the underlying clusters
+and then min replicas are distributed to those clusters which got max. This means
+that each cluster will get an hpa if the max replicas specified are greater then
```
are greater then -> is greater than

tengqm commented on docs/tasks/administer-federation/hpa.md:
```diff
@@ -0,0 +1,154 @@
+and then min replicas are distributed to those clusters which got max. This means
+that each cluster will get an hpa if the max replicas specified are greater then
+the total clusters participating in this federation and some clusters will be
+skipped if max replicas specified are lesser then the total clusters participating
```
are lesser then -> is less than

tengqm commented on docs/tasks/administer-federation/hpa.md:
```diff
@@ -0,0 +1,154 @@
+
+To a `federation user` interacting with federated control plane (or simply federation),
+the interaction is same as interacting with a normal k8s cluster (but with a limited set of
+apis, that which are federated). As both deployments and horizontalpodautoscalers are
```
apis, that which are federated -> APIs that are federated

irfanurrehman commented:
I have updated the doc to fix the review comments. 

---

Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/tasks/administer-federation/hpa.md, line 2 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6cP4ieV8S1Ffy2fHv:-KuGR4xudt1oAaua584d:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L2)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary><blockquote>

I'd spell it out: "Federated Horizontal Pod Autoscalers (HPA)"
</blockquote></details>

Done.

---

*[docs/tasks/administer-federation/hpa.md, line 6 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6cZXZ7ZAJPl8CDq-k:-KuGR7DuepOpHfSuwyTp:b7a6sb3) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L6)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary><blockquote>

HPA's ?
</blockquote></details>

made hpa -> HPA at all places in the doc.

---

*[docs/tasks/administer-federation/hpa.md, line 65 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6d8lyDnMx1VQpc-ZX:-KuGRFn2pA8mBo6Q7lK-:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L65)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary><blockquote>

Please mention cluster selectors to restrict the clusters into which the HPA gets federated.
</blockquote></details>

Done.

---

*[docs/tasks/administer-federation/hpa.md, line 79 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku64V96UCpmouTO9Nup-r1-79:-KuGRGwhSKDPDHr-Z5FF:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L79)):*
<details><summary><i>Previously, tengqm (Qiming) wrote…</i></summary><blockquote>

then -> than
</blockquote></details>

Done.

---

*[docs/tasks/administer-federation/hpa.md, line 80 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku64V96UCpmouTO9Nup-r1-80:-KuGRIefEwdF1sfRvIjv:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L80)):*
<details><summary><i>Previously, tengqm (Qiming) wrote…</i></summary><blockquote>

greater -> greater than
</blockquote></details>

Done.

---

*[docs/tasks/administer-federation/hpa.md, line 88 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku64V96UCpmouTO9Nup-r1-88:-KuGRJeDhNf1nBC9zu4c:b8wxksg) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L88)):*
<details><summary><i>Previously, tengqm (Qiming) wrote…</i></summary><blockquote>

are greater then -> is greater than
</blockquote></details>

Updated the sentence, please check if still sounds wrong.

---

*[docs/tasks/administer-federation/hpa.md, line 90 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku64V96UCpmouTO9Nup-r1-90:-KuGRRKLD2cHg9hot588:b8wxksg) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L90)):*
<details><summary><i>Previously, tengqm (Qiming) wrote…</i></summary><blockquote>

are lesser then -> is less than
</blockquote></details>

Updated the sentence, please check if still sounds wrong.

---

*[docs/tasks/administer-federation/hpa.md, line 118 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6dlHCXQY6RbnIv8n2:-KuGRTnCp9r9QM_nQXZM:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L118)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary><blockquote>

I think it's worth mentioning that this will delete all underlying HPA's, if cascading deletion is used.
</blockquote></details>

Done.

---

*[docs/tasks/administer-federation/hpa.md, line 131 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6dzROL-7xwYkv-Jyr:-KuGRUglAqW8k8VYNCAw:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L131)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary><blockquote>

'almost identical' rather than 'the same'.
</blockquote></details>

Done.

---

*[docs/tasks/administer-federation/hpa.md, line 132 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku64V96UCpmouTO9Nup-r1-132:-KuGRVXwjODTrYBralwY:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L132)):*
<details><summary><i>Previously, tengqm (Qiming) wrote…</i></summary><blockquote>

apis, that which are federated -> APIs that are federated
</blockquote></details>

Done.

---

*[docs/tasks/administer-federation/hpa.md, line 146 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-Ku6eInOhNVJSQjqNgY8:-KuGRWt8x9OqpCkdsqZo:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/917ae497eb614e35c17c749bbb0ef9c5e04fe30c/docs/tasks/administer-federation/hpa.md#L146)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary><blockquote>

I think you need to be clearer here that the federation does not actually monitor the CPU utilization itself, but rather relies on the underlying cluster HPA's to do the auto-scaling, and only re-assigns min and max replicas between clusters in an attempt to prevent individual clusters from running up against their max replica limits.  A brief example might help to clarify this.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-:-KuGR_Iil2PPLxeNbrGH:b81m1fr)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
@quinton-hoole @zacharysarah
I still think this PR should wait for https://github.com/kubernetes/kubernetes/pull/49950 to get merged.
I believe that would happen when the code freeze is lifted.

quinton-hoole commented:
The PR  kubernetes/kubernetes#49950 has merged.  All edits /lgtm.  This can be merged.

steveperry-53 commented:
I'll start a docs review now.

steveperry-53 commented:
@irfanurrehman See https://github.com/kubernetes/kubernetes.github.io/pull/5487/commits/30df752ffc61f8c0a34e8100676f094ef8038d8a for suggested edits.

steveperry-53 commented:
@irfanurrehman Ping. As far as I know, we're still shooting for a 9/27 release.

quinton-hoole commented:
Thanks @steveperry-53 .  Yes, still aiming at 9/27 release.  Perhaps @irfanurrehman was unaware that he was responsible for applying your edits. Is the documentation process that we should follow documented somewhere?  Any pointers much appreciated.

zacharysarah commented:
@quinton-hoole :wave: 

> Perhaps @irfanurrehman was unaware that he was responsible for applying your edits. 

Generally, docs contributors provide reviews and perform small edits.* In the case of more extensive edits, docs contributors provide feedback for a PR owner to implement. In the case of this PR specifically, I edited https://github.com/kubernetes/kubernetes.github.io/commit/30df752ffc61f8c0a34e8100676f094ef8038d8a last week; Steve's feedback 

> Is the documentation process that we should follow documented somewhere?

Not in terms of PR review. Looking at the [README](https://github.com/kubernetes/kubernetes.github.io) and the [contribution guidelines](https://kubernetes.io/docs/home/contribute/create-pull-request/), there's definitely room to set clearer expectations. I've opened an issue to improve the review process: https://github.com/kubernetes/kubernetes.github.io/issues/5620

For now, we need either:
- @irfanurrehman to push a commit that incorporates updates
- @irfanurrehman to un-check and re-check the box allowing edits from maintainers

[*]The GitHub UI has a known issue: edits from maintainers are allowed only if the PR owner un-checks and re-checks the "allow edits from maintainers" box after the PR is created.

steveperry-53 commented:
@quinton-hoole @zacharysarah @irfanurrehman, We don't need anyone to check the **Allow edits** box, and we don't need @irfanurrehman to push anything. I already pushed a commit with suggested edits. All we need is for someone to agree to what I've done in https://github.com/kubernetes/kubernetes.github.io/pull/5487/commits/30df752ffc61f8c0a34e8100676f094ef8038d8a.

quinton-hoole commented:
@steveperry-53 Yes, thanks, https://github.com/kubernetes/kubernetes.github.io/commit/30df752ffc61f8c0a34e8100676f094ef8038d8a looks fine. 

/lgtm

Let me know if you need us to do anything else.

steveperry-53 commented:
Reviewed 2 of 3 files at r1, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487)*
<!-- Sent from Reviewable.io -->

steveperry-53 commented:
@quinton-hoole Thanks for the lgtm.

@quinton-hoole, @zacharysarah Can one of you help get the code-review/reviewable to be green. To this day, I am mystified by Reviewable. Thanks.

steveperry-53 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/5487#-:-KuvYhB_EkBLTv5L95BC:bnfp4nl)*
<!-- Sent from Reviewable.io -->
